### PR TITLE
(closes #2419) Rename UnknownType to UnsupportedType and DeferredType to UnresolvedType

### LINKED ISSUE #2419: Rename "unknown" symbol types?
Currently we have:
UnknownType - We known the Fortran string but we don't know what it means.
DeferredType - We don't know the Fortran string, but maybe we could find it out by resolving the modules.

It will be clearer if we renamed:
UnknownType -> UnsupportedType (as it makes clear that the type fortran string in known)
DeferredType -> UnresolvedType (as it doesn't clash with the "deferred" Fortran type attribute and it uses the same word as UnresolvedInterface - when we don't know the symbol comes from)

### PR DESCRIPTION
**Pull request by sergisiso**

### CODE REVIEW

**Comment by sergisiso** [comment]
This is ready to review, one for @arporter or @hiker

**Comment by arporter** [comment]
I managed to find:
 
   $ find ./doc ./src -name \*.py -exec grep -iH unknownfortrantype {} \;
    ./src/psyclone/tests/psyir/frontend/fparser2_test.py:def test_process_declarations_unknownfortrantype():
    ./src/psyclone/psyir/backend/fortran.py:    (This is required because we capture an UnknownFortranType declaration as

**Review by arporter — CHANGES_REQUESTED** [review]
Good work Sergi. Just a couple of minor things and one pycodestyle issue:

    psyclone/tests/psyir/nodes/array_reference_test.py:550:62: E128 continuation line under-indented for visual indent

I've fired off the integration tests now.

File: src/psyclone/tests/psyir/nodes/operation_test.py [hunk 1]
```diff
@@ -187,14 +187,14 @@ def test_binaryop_scalar_datatype():
...
+    # When any one of the arguments is of UnsupportedType then we know
     # nothing about the type of the result.
     uref1 = Reference(
         DataSymbol("trouble",
-                   UnknownFortranType("real, volatile :: trouble")))
+                   UnsupportedFortranType("real, volatile :: trouble")))
     binop4 = BinaryOperation.create(oper, iref1.copy(), uref1)
-    # TODO #2419 - DeferredType should probably be UnsupportedType really.
-    assert isinstance(binop4.datatype, DeferredType)
+    # TODO #2419 - UnresolvedType should probably be UnsupportedType really.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/nodes/operation_test.py:196]: yep, done

File: src/psyclone/tests/psyir/backend/opencl_test.py [hunk 2]
```diff
@@ -247,7 +247,7 @@ def data_arguments(self):
         "  return;\n" \
         "}\n\n"
 
-    # Add a symbol with a deferred interface and check that this raises the
+    # Add a symbol with a Unresol and check that this raises the
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/opencl_test.py:250]: Ups. Done.

File: src/psyclone/tests/psyir/backend/fortran_unknown_declns_test.py [hunk 3]
```diff
@@ -110,14 +110,14 @@ def test_fw_unknown_interface_decln(tmpdir, fortran_writer):
     assert Compile(tmpdir).string_compiles(fortran_writer(container))
 
 
-def test_fw_unknowntype_routine_symbols_error(fortran_writer):
+def test_fw_unsupportedtype_routine_symbols_error(fortran_writer):
     ''' Check that the backend raises the expected error if a RoutineSymbol
-    which is not imported or an interface UnknownFortanType is found by the
+    which is not imported or an interface UnsupportedFortanType is found by the
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/fortran_unknown_declns_test.py:115]: Rephrased to "or a Fortran interface (currently an UnsupportedFortanType)" which is what I think we mean

File: src/psyclone/psyir/symbols/typed_symbol.py [hunk 4]
```diff
@@ -150,8 +150,8 @@ def copy_properties(self, symbol_in):
         super(TypedSymbol, self).copy_properties(symbol_in)
         self._datatype = symbol_in.datatype
 
-    def resolve_deferred(self):
-        ''' If the symbol has a deferred datatype, find where it is defined
+    def resolve_type(self):
+        ''' If the symbol has am Unresolved datatype, find where it is defined
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/symbols/typed_symbol.py:154]: Fixed

**Comment by sergisiso** [comment]
> I managed to find:

$ find ./doc ./src -name *.py -exec grep -iH unknownfortrantype {} ;
./src/psyclone/tests/psyir/frontend/fparser2_test.py:def test_process_declarations_unknownfortrantype():
./src/psyclone/psyir/backend/fortran.py: (This is required because we capture an UnknownFortranType declaration as

Done!, which also made me think of doing "find . -name *unknown*" and found one filename to update.

> I've fired off the integration tests now.

I fired it two days ago in the same commit. I just saying it because I don't think the CI repeats the test to the same commit and you will have to go back in the jobs list to find the results, which were successful.

**Comment by arporter** [comment]
> I fired it two days ago in the same commit. I just saying it because I don't think the CI repeats the test to the same commit and you will have to go back in the jobs list to find the results, which were successful.

Oh, that's clever, I didn't know it did that. It also explains why I can see no trace of the job I submitted half an hour ago!

File: src/psyclone/psyir/frontend/fparser2.py [hunk 5]
```diff
@@ -3980,7 +3980,7 @@ def _data_ref_handler(self, node, parent):
             # Base of reference is a scalar entity and must be a DataSymbol.
             base_sym = _find_or_create_unresolved_symbol(
                 parent, node.children[0].string.lower(),
-                symbol_type=DataSymbol, datatype=DeferredType())
+                symbol_type=DataSymbol, datatype=UnresolvedType())
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/frontend/fparser2.py:3983]: Note that now we have Unresolved Symbols (interfaces) and Unresolved DataTypes (and both have the same meaning - that the declaration has not been found), but in this case we mean the former (although in many cases the later would also be true). As you mention, I think I would like to leave this for another PR.

File: src/psyclone/psyir/symbols/datatypes.py [hunk 6]
```diff
@@ -68,17 +68,12 @@ def __eq__(self, other):
...
+    ''' Indicates that the type declaration has not been found yet. '''
 
     def __str__(self):
-        '''
-        :returns: a description of this deferred type.
-        :rtype: str
-
-        '''
-        return "DeferredType"
+        return "UnresolvedType"
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/symbols/datatypes.py:76]: Again, as you suggest I prefer not changing other aspects in this PR as its big enough.

**Comment by sergisiso** [comment]
@arporter This is ready for the next review.

**Review by arporter — APPROVED** [review]
All good now thanks.
I've also checked for the issue number and not found any mentions of it.
All the different docs build fine.
Will proceed to merge.

**Pull request merged**